pytket/pytket/utils/distribution.py

```diff
@@ -0,0 +1,268 @@
+        """Combine two distributions."""
+        return EmpiricalDistribution(self._C + other._C)
+
+    def condition(self, criterion: Callable[[Any], bool]) -> "EmpiricalDistribution":
+        """Return a new distribution conditioned on the given criterion.
+
+        :param criterion: A boolean function defined on all possible observations.
+        """
+        return EmpiricalDistribution(
+            Counter({x: c for x, c in self._C.items() if criterion(x)})
+        )
+
+    def map(self, mapping: Callable[[Any], Any]) -> "EmpiricalDistribution":
+        """Return a distribution over a transformed domain.
+
+        The provided function maps elements in the original domain to new elements. If
+        it is not injective, counts are combined.
+
+        :param mapping: A function defined on all possible observations, mapping them
+            to another domain.
+        """
+        C: Counter = Counter()
+        for x, c in self._C.items():
+            C[mapping(x)] += c
+        return EmpiricalDistribution(C)
+
+    def sample_mean(self, f: Callable[[Any], Number]) -> Number:
+        """Compute the sample mean of a functional.
+
+        The provided function maps observations to numerical values.
+
+        :return: Estimate of the mean of the functional based on the observations."""
+        return sum(c * f(x) for x, c in self._C.items()) / self.total
+
+    def sample_variance(self, f: Callable[[Any], Number]) -> Number:
+        """Compute the sample variance of a functional.
+
+        The provided function maps observations to numerical values.
+
+        The sample variance is an unbiased estimate of the variance of the underlying
+        distribution.
+
+        :return: Estimate of the variance of the functional based on the
+            observations."""
+        if self.total < 2:
+            raise RuntimeError(
+                "At least two samples are required in order to compute the sample "
+                "variance."
+            )
+        fs = [(f(x), c) for x, c in self._C.items()]
+        M0 = self.total
+        M1 = sum(c * v for v, c in fs)
+        M2 = sum(c * v**2 for v, c in fs)
+        return (M2 - M1**2 / M0) / (M0 - 1)
+
+
+class ProbabilityDistribution:
+    """Represents an exact probability distribution.
+
+    Supports methods for combination, marginalization, expectation value, etc. May be
+    derived from an :py:class:`EmpriricalDistribution`.
+    """
+
+    def __init__(self, P: Dict[Any, float]):
+        """Initialize with a dictionary of probabilities.
+
+        The values must be non-negative and add up to at most 1.
+        """
+        if any(x < 0 for x in P.values()):
+            raise ValueError("Distribution contains negative probabilities")
+        S = sum(P.values())
+        if not np.isclose(S, 1):
+            raise ValueError("Probabilities do not sum to 1")
+        self._P: Dict[Any, float] = {x: p for x, p in P.items() if not np.isclose(p, 0)}
```

Comment:
Could the `isclose` to 1 check for the sum fail if all the `isclose` to zero values are removed?

Reply:
In theory, yes. We could allow construction from non-normalized values (with scaling applied in the constructor), to get around this.

Reply:
Fixed,